src/Receiver.cc

```diff
@@ -606,32 +415,44 @@ Receiver::checkResendTimeouts(uint64_t now, MessageBucket* bucket)
         return;
     }
 
+    SpinLock::Lock lock_bucket(bucket->mutex);
     while (true) {
-        SpinLock::Lock lock_bucket(bucket->mutex);
-
         // No remaining timeouts.
         if (bucket->resendTimeouts.empty()) {
             break;
         }
 
-        Message* message = &bucket->resendTimeouts.front();
-
         // No remaining expired timeouts.
+        Message* message = &bucket->resendTimeouts.front();
         if (!message->resendTimeout.hasElapsed(now)) {
             break;
         }
 
         // Found expired timeout.
-        assert(message->state == Message::State::IN_PROGRESS);
-        bucket->resendTimeouts.setTimeout(&message->resendTimeout);
+        assert(message->getState() == Message::State::IN_PROGRESS);
+        message->numResendTimeouts++;
+        if (message->numResendTimeouts >= MESSAGE_TIMEOUT_INTERVALS) {
+            // Message timed out before being fully received; drop the message.
+            messageAllocator.destroy(message);
```

Comment:
Does this deadlock on `~Message()`? The `bucket->mutex` is held and also used the destructor, correct?

Reply:
Yeah, good catch. Now I used a `UniqueLock` and released the lock before calling `dropMessage` (the alternative is to duplicate functionality from `dropMessage` like the original code). Yes, I also revived `dropMessage` because it feels like information leakage to detach a message from the transport inside its destructor.